src/i18n-helpers.js

```diff
@@ -0,0 +1,154 @@
+      this.saveCaches();
+    }, 5_000);
+  }
+
+  fetchAST(ns, _path, str, lang) {
+    let path = [ns, lang, ..._path];
+
+    let ast = optionalTravelDownObjectPath(
+      this.cachedASTs,
+      path
+    );
+    if (ast && "_AST" in ast) return ast._AST;
+
+    ast = parse(str, {
+      // requiresOtherClause
+    });
+
+    let lastBit = path.pop();
+    let cachePath = travelDownOrMakePath(this.cachedASTs, path);
+    cachePath[lastBit] = { _AST: ast };
+
+    this.debouncedCleanAndSave();
+
+    return ast;
+  }
+
+  /**
+   * go through `this.cachedASTs`, find stuff that doesn't exist in `registeredStrings`,
+   * then yeet them
+   */
+  cleanCaches(registeredStrings, cachedASTs) {
+    if (!cachedASTs) cachedASTs = this.cachedASTs;
+
+    Object.entries(cachedASTs).forEach(([k, cachedValue]) => {
+      let registeredValue = registeredStrings[k];
+
+      // path doesn't exist
+      if (!registeredValue) {
+        delete cachedASTs[k];
+        return;
+      }
+
+      // path is an object
+      if (typeof registeredValue === "object") {
+        // cached is not AST (plain obj) (good)
+        if (
+          typeof cachedValue === "object"
+          && !("_AST" in cachedValue)
+        ) {
+          this.cleanCaches(registeredValue, cachedValue);
+          return;
+        }
+        // cached is AST (bad)
+        delete cachedASTs[k];
+        return;
+      }
+
+      // path is a string
+      if (typeof registeredValue === "string") {
+        // cached is AST (good)
+        if ("_AST" in cachedValue) return;
+        // cached is not AST (bad)
+        delete cachedASTs[k];
+        return;
+      }
+    });
+  }
+
+  saveCaches() {
+    let cachedir = path.join(
+      this.configDirPath,
+      "compile-cache",
+      "i18n"
+    );
+    fs.mkdirSync(cachedir, { recursive: true });
+
+    let cachefile = path.join(cachedir, "strings.json");
+    fs.writeFileSync(cachefile, JSON.stringify(this.cachedASTs));
+  }
+
+  loadCaches() {
+    let cachefile = path.join(
+      this.configDirPath,
+      "compile-cache",
+      "i18n",
+      "strings.json"
+    );
+    if (fs.existsSync(cachefile)) {
+      this.cachedASTs = JSON.parse(fs.readFileSync(cachefile, "utf-8"));
+    }
+  }
+}
+
+function walkStrings(strings, cb, accum = []) {
+  Object.entries(strings).forEach(([k, v]) => {
+    let path = [...accum, k];
+    if (typeof v === "string") cb(path, v, true);
+    else if (typeof v === "object") {
+      cb(path, null, false);
+      walkStrings(v, cb, path);
+    }
+  });
+}
+
+function travelDownObjectPath(obj, path) {
+  for (const pathFragment of path) {
+    obj = obj[pathFragment];
+  }
+  return obj;
+}
+
+function optionalTravelDownObjectPath(obj, path) {
+  for (const pathFragment of path) {
+    obj = obj[pathFragment];
+    if (!obj) return undefined;
+  }
+  return obj;
+}
+
+function travelDownOrMakePath(obj, path) {
+  for (const pathFragment of path) {
```

Comment:
```suggestion
function travelDownOrMakePath(obj, path) {
  if (!Array.isArray(path)) { return undefined; }
  for (const pathFragment of path) {
```

The above is just a suggestion for these functions all above. 
Returning `undefined` isn't important, could be `false` or `0` or whatever you think makes the most sense.
Just with these being so close to being part of a public API any package could hit I think there ought to be some importance in catching the errors that could cause the editor to crash. Since obviously we want to avoid it as much as we can, which this could protect against. But if this is confidently handled elsewhere fair enough, just to me without knowing the ins and out here seemed like a good thing to mention.

Reply:
I made this function as an internal helper, so I should be the only one calling it, and I did make sure that they are all passing in arrays

Reply:
Okay super awesome. Then yeah nothing to worry about here as long as it's protected somewhere in the call stack.